external_compressors/zstd_external.py

```diff
@@ -0,0 +1,170 @@
+        with Uint8FileDataStream(input_file_path, "rb") as fds:
+            with EncodedBlockWriter(encoded_file_path) as writer:
+                self.encode(fds, block_size=block_size, encode_writer=writer)
+
+
+class ZstdExternalDecoder(DataDecoder):
+    def __init__(self, level=6):
+        self.level = level
+        self.zstd_obj = create_zstd_decompressor_object()
+        self.block_size_num_bits = 32  # num used to encode the block size at start
+
+    def reset(self):
+        self.zstd_obj = create_zstd_decompressor_object()
+
+    def decode_block(self, compressed_bitarray: BitArray):
+        # first read the size of the zstd block(s)
+        zstd_size = bitarray_to_uint(compressed_bitarray[: self.block_size_num_bits])
+        zstd_compressed = compressed_bitarray[
+            self.block_size_num_bits : self.block_size_num_bits + zstd_size
+        ]
+
+        compressed_bytes = zstd_compressed.tobytes()
+        return (
+            DataBlock(list(self.zstd_obj.decompress(compressed_bytes))),
+            self.block_size_num_bits + zstd_size,
+        )
+
+    def decode_file(self, encoded_file_path: str, output_file_path: str):
+        """utility wrapper around the decode function using Uint8FileDataStream
+
+        Args:
+            encoded_file_path (str): input binary file
+            output_file_path (str): output (text) file to which decoded data is written
+        """
+
+        # decode data and write output to a text file
+        with EncodedBlockReader(encoded_file_path) as reader:
+            with Uint8FileDataStream(output_file_path, "wb") as fds:
+                self.decode(reader, fds)
+
+
+def test_zstd_encode_decode():
+    encoder = ZstdExternalEncoder()
+    decoder = ZstdExternalDecoder()
+
+    # create some sample data consisting of bytes
+    data_list = [0, 0, 1, 3, 4, 100, 255, 123, 234, 42, 186]
+    data_block = DataBlock(data_list)
+
+    is_lossless, _, _ = try_lossless_compression(
+        data_block, encoder, decoder, add_extra_bits_to_encoder_output=True
+    )
+    assert is_lossless
+
+
+def test_zstd_file_encode_decode():
+    """full test for ZstdExternalEncoder and ZstdExternalDecoder
+
+    - create a sample file
+    - encode the file using ZstdExternalEncoder
+    - perform decoding and check if the compression was lossless
+
+    """
+    # define encoder, decoder
+    encoder = ZstdExternalEncoder()
+    decoder = ZstdExternalDecoder()
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # create a file with some random data
+        input_file_path = os.path.join(tmpdirname, "inp_file.txt")
+        create_random_binary_file(
+            input_file_path,
+            file_size=5000,
+            prob_dist=ProbabilityDist({44: 0.5, 45: 0.25, 46: 0.2, 255: 0.05}),
+        )
+
+        # test lossless compression
+        assert try_file_lossless_compression(
+            input_file_path, encoder, decoder, encode_block_size=1000
+        )
```

Comment:
Should we also add a test case to check that the encoded file size matches the one generated via Zstandard if it was run stand-alone outside SCL?